File: frontend/src/services/notificationService.js
import api from "./api";

const notificationService = {
  // ==========================
  // Get All Notifications
  // ==========================
  async getNotifications() {
    try {
      const response = await api.get("/notifications");
      return response.data;
    } catch (error) {
      console.error("Error fetching notifications:", error);
      throw error;
    }
  },

  // ==========================
  // Get Unread Notifications
  // ==========================
  async getUnreadNotifications() {
    try {
      const response = await api.get("/notifications/unread");
      return response.data;
    } catch (error) {
      console.error("Error fetching unread notifications:", error);
      throw error;
    }
  },

  // ==========================
  // Get Notification By ID
  // ==========================
  async getNotificationById(id) {
    try {
      const response = await api.get(`/notifications/${id}`);
      return response.data;
    } catch (error) {
      console.error("Error fetching notification:", error);
      throw error;
    }
  },

  // ==========================
  // Mark Single Notification Read
  // ==========================
  async markAsRead(id) {
    try {
      const response = await api.patch(
        `/notifications/${id}/read`
      );

      return response.data;
    } catch (error) {
      console.error("Error marking notification as read:", error);
      throw error;
    }
  },

  // ==========================
  // Mark All Notifications Read
  // ==========================
  async markAllAsRead() {
    try {
      const response = await api.patch(
        "/notifications/read-all"
      );

      return response.data;
    } catch (error) {
      console.error("Error marking all notifications as read:", error);
      throw error;
    }
  },

  // ==========================
  // Delete Notification
  // ==========================
  async deleteNotification(id) {
    try {
      const response = await api.delete(
        `/notifications/${id}`
      );

      return response.data;
    } catch (error) {
      console.error("Error deleting notification:", error);
      throw error;
    }
  },

  // ==========================
  // Clear All Notifications
  // ==========================
  async clearAllNotifications() {
    try {
      const response = await api.delete(
        "/notifications"
      );

      return response.data;
    } catch (error) {
      console.error("Error clearing notifications:", error);
      throw error;
    }
  },
};

export default notificationService;

File: backend/app/services/notification_service.py
from typing import List, Optional
from sqlalchemy.orm import Session

from app.models.notification import Notification


def create_notification(
    db: Session,
    title: str,
    message: str,
    type: str = "info",
    truck_id: Optional[str] = None,
) -> Notification:
    """
    Create and save a new notification.
    """
    notification = Notification(
        title=title,
        message=message,
        type=type,
        is_read=False,
        truck_id=truck_id,
    )
    db.add(notification)
    db.commit()
    db.refresh(notification)
    return notification


def get_notifications(db: Session, limit: int = 100) -> List[Notification]:
    """
    Get all notifications ordered by most recent.
    """
    return (
        db.query(Notification)
        .order_by(Notification.created_at.desc())
        .limit(limit)
        .all()
    )


def get_unread_notifications(db: Session) -> List[Notification]:
    """
    Get unread notifications.
    """
    return (
        db.query(Notification)
        .filter(Notification.is_read.is_(False))
        .order_by(Notification.created_at.desc())
        .all()
    )


def get_notification(db: Session, notification_id: int) -> Optional[Notification]:
    """
    Get notification by ID.
    """
    return (
        db.query(Notification)
        .filter(Notification.id == notification_id)
        .first()
    )


def mark_as_read(db: Session, notification_id: int) -> Optional[Notification]:
    """
    Mark single notification as read.
    """
    notification = get_notification(db, notification_id)
    if not notification:
        return None

    notification.is_read = True
    db.commit()
    db.refresh(notification)
    return notification


def mark_all_as_read(db: Session) -> int:
    """
    Mark all unread notifications as read.
    """
    count = (
        db.query(Notification)
        .filter(Notification.is_read.is_(False))
        .update({"is_read": True})
    )
    db.commit()
    return count


def delete_notification(db: Session, notification_id: int) -> bool:
    """
    Delete notification by ID.
    """
    notification = get_notification(db, notification_id)
    if not notification:
        return False

    db.delete(notification)
    db.commit()
    return True


def clear_all_notifications(db: Session) -> int:
    """
    Delete all notifications.
    """
    count = db.query(Notification).delete()
    db.commit()
    return count
